src/web/csp.rs

```diff
@@ -0,0 +1,189 @@
+use crate::config::Config;
+use iron::{AfterMiddleware, BeforeMiddleware, IronResult, Request, Response};
+
+pub(super) struct Csp {
+    nonce: String,
+    suppress: bool,
+}
+
+impl Csp {
+    fn new() -> Self {
+        let mut random = [0u8; 36];
+        getrandom::getrandom(&mut random).expect("failed to generate a nonce");
+        Self {
+            nonce: base64::encode(&random),
+            suppress: false,
+        }
+    }
+
+    pub(super) fn suppress(&mut self, suppress: bool) {
+        self.suppress = suppress;
+    }
+
+    pub(super) fn nonce(&self) -> &str {
+        &self.nonce
+    }
+
+    fn render(&self, content_type: ContentType) -> Option<String> {
+        if self.suppress {
+            return None;
+        }
+        let mut result = String::new();
+
+        // Disable everything by default
+        result.push_str("default-src 'none'");
+
+        // Disable the <base> HTML tag to prevent injected HTML content from changing the base URL
+        // of all relative links included in the website.
+        result.push_str("; base-uri 'none'");
+
+        // Allow loading images from the same origin. This is added to every response regardless of
+        // the MIME type to allow loading favicons.
+        //
+        // Images from other HTTPS origins are also temporary allowed until issue #66 is fixed.
+        result.push_str("; img-src 'self' https:");
+
+        match content_type {
+            ContentType::Html => self.render_html(&mut result),
+            ContentType::Svg => self.render_svg(&mut result),
+            ContentType::Other => {}
+        }
+
+        Some(result)
+    }
+
+    fn render_html(&self, result: &mut String) {
+        // Allow loading any CSS file from the current origin.
+        result.push_str("; style-src 'self'");
+
+        // Allow loading any font from the current origin.
+        result.push_str("; font-src 'self'");
+
+        // Only allow scripts with the random nonce attached to them.
+        //
+        // We can't just allow 'self' here, as users can upload arbitrary .js files as part of
+        // their documentation and 'self' would allow their execution. Instead, every allowed
+        // script must include the random nonce in it, which an attacker is not able to guess.
```

Comment:
Ahhh, I just realized while reading this that there's a new nonce *for each request*, it's not a hash of the file or anything like that. Maybe add a comment to the CSP saying that? And a test that two sequential requests get different nonces?

Reply:
[Hmm, the test is already there](https://github.com/rust-lang/docs.rs/pull/1333/files#diff-6f288f9159b4a7665a0894ea5bbfb4de699bebe2de84a24b423895e0e0daae09R138-R143). I'll add a comment.